Rewrite new_zeros, new_ones, new_full decomp with aten.full

SherlockNoMad opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #84946

We should **NOT**  introducing non-functional op for decomps of functional op. 

For example 
```
make_fx(functionalize(lambda x: x.new_zeros(3)), decomposition_table=decomposition_table)(x)
```
is producing 
```
def forward(self, x_1):
    empty = torch.ops.aten.empty.memory_format([3, 4], dtype = torch.float32, layout = torch.strided, device = device(type='cpu'), pin_memory = False)
    zero_ = torch.ops.aten.zero_.default(empty);  empty = None
    return zero_
```

SherlockNoMad committed:
Rewrite new_zeros, new_ones, new_full decomp with aten.full

[ghstack-poisoned]

SherlockNoMad commented:
We can further extend this to ones, zeros...

ngimel commented:
cc @Lezcano, @mruberry

ngimel commented on torch/_refs/__init__.py:
```diff
@@ -3320,11 +3316,9 @@ def new_full(
-    r = a.new_empty(
-        size, dtype=dtype, layout=layout, device=device, pin_memory=pin_memory
+    return torch.full(
+        size, fill_value, dtype=dtype, layout=layout, device=device, pin_memory=pin_memory
```
you are incorrectly setting dtype and device - in case they are not supplied, they are taken from `a`. Test errors are real.

SherlockNoMad replied:
fixed.

SherlockNoMad committed:
Update on "Rewrite new_zeros, new_ones, new_full decomp with aten.full"


We should **NOT**  introducing non-functional op for decomps of functional op. 

For example 
```
make_fx(functionalize(lambda x: x.new_zeros(3)), decomposition_table=decomposition_table)(x)
```
is producing 
```
def forward(self, x_1):
    empty = torch.ops.aten.empty.memory_format([3, 4], dtype = torch.float32, layout = torch.strided, device = device(type='cpu'), pin_memory = False)
    zero_ = torch.ops.aten.zero_.default(empty);  empty = None
    return zero_
```




[ghstack-poisoned]

SherlockNoMad committed:
Update on "Rewrite new_zeros, new_ones, new_full decomp with aten.full"


We should **NOT**  introducing non-functional op for decomps of functional op. 

For example 
```
make_fx(functionalize(lambda x: x.new_zeros(3)), decomposition_table=decomposition_table)(x)
```
is producing 
```
def forward(self, x_1):
    empty = torch.ops.aten.empty.memory_format([3, 4], dtype = torch.float32, layout = torch.strided, device = device(type='cpu'), pin_memory = False)
    zero_ = torch.ops.aten.zero_.default(empty);  empty = None
    return zero_
```




[ghstack-poisoned]

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -3275,6 +3275,20 @@ def new_empty_strided(
 
 
+@register_decomposition(torch.ops.aten.zeros)
+def zeros(
```
What about the `out` parameter?

SherlockNoMad replied:
This is the decomp for functional version of zero.
For out-variant ops, I think we need a separated decomp func.

Lezcano replied:
Can't we just use the `out` wrapper? It should register the correct decomposition as well.

SherlockNoMad replied:
ic what you mean... added.

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -3275,6 +3275,20 @@ def new_empty_strided(
+
+
 @register_decomposition(torch.ops.aten.new_zeros)
 def new_zeros(
```
What about the `requires_grad` parameter?

mruberry replied:
(If the ATen version doesn't support `requires_grad` but the torch version does then let's at least add a comment to that effect -- we'll probably want to separate references that are different between ATen and torch soon)

SherlockNoMad replied:
added.

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -3285,11 +3299,27 @@ def new_zeros(
-    r = a.new_empty(
-        size, dtype=dtype, layout=layout, device=device, pin_memory=pin_memory
+    dtype = a.dtype if dtype is None else dtype
+    layout = a.layout if layout is None else layout
```
I think the documentation for `new_zeros` it out-of-date:

https://pytorch.org/docs/master/generated/torch.Tensor.new_zeros.html?highlight=new_zeros#torch.Tensor.new_zeros

It doesn't even refer to layout or pinning memory

SherlockNoMad replied:
Tracked with https://github.com/pytorch/pytorch/issues/84986

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -3275,6 +3275,20 @@ def new_empty_strided(
+    pin_memory: bool = False,
+) -> TensorLikeType:
+    return torch.full(
+        size, 0, dtype=dtype, layout=layout, device=device, pin_memory=pin_memory
```
How does this work? 

```
torch.zeros((3, 3)).dtype
: torch.float32

torch.full((3, 3), 0).dtype
: torch.int64
```

Does the dtype passed to `full` need to be the default floating point dtype instead of `None`?

mruberry replied:
Also, it's an interesting quirk that we permit integer->bool Python type casts for numbers

Lezcano replied:
Not only we allow `integer->bool` Python casts, but people have used them quite a bit throughout refs. This is not my cup of tea, but I guess it makes the code slightly more compact.

SherlockNoMad replied:
See updated code...

zeros' ref is using fill_value of False, ones' ref is  fill_value of True. 
Values of False and True can be casted to float or int, by just setting the expected dtype in torch.full.

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -3302,29 +3332,38 @@ def new_ones(
     a: TensorLikeType,
     size: ShapeType,
-    fill_value: NumberType,
+    fill_value: Union[int, float, bool],
```
We should cite an issue for this in a comment. Seems like `new_full()` should allow complex fill values.

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -3771,17 +3810,25 @@ def eye(
+    requires_grad: bool = False,
 ) -> TensorLikeType:
-    e = empty(shape, dtype=dtype, device=device, requires_grad=requires_grad)
+    e = empty(
```
Doesn't the dtype passed to `empty()` need to be inferred from the `fill_value` type?

ngimel replied:
pre-existing condition

SherlockNoMad replied:
torch.full is respecting the dtype value, instead of inferring from the fill_value's dtype.

```
>>> torch.full((3,), False, dtype=torch.float)
tensor([0., 0., 0.])
```

ngimel replied:
The question here was what to do when dtype is not passed. `full` infers dtype from fill value, whereas `empty` would create default dtype tensor when dtype is None (and thus `full` result will be wrong)

SherlockNoMad replied:
ic... 
but why we don't we directly use prims.full? Is there any particular reason we are doing empty+fill?

ngimel replied:
we could, but device/dtype etc args in prims.full are non-optional, so they still need to be set (btw @mruberry we probably need to assert that those args are set in prims?)

mruberry replied:
They are required args when calling the prim:

https://github.com/pytorch/pytorch/blob/4db1588ca09b4f6328c5bd98701e9c302cb39800/torch/_prims/__init__.py#L2357

Using `full` directly sounds better than calling `empty` and then `fill` in most cases, there are some caveats with modeling new functions that we may or may not be able to model now, however. The point @ngimel and I are making is that the metadata for the `prims.full` call needs to be inferred by the reference.

mruberry commented:
Hey @SherlockNoMad! I made some comments inline for your review. Looking forward to hearing your thoughts

SherlockNoMad committed:
Update on "Rewrite new_zeros, new_ones, new_full decomp with aten.full"


We should **NOT**  introducing non-functional op for decomps of functional op. 

For example 
```
make_fx(functionalize(lambda x: x.new_zeros(3)), decomposition_table=decomposition_table)(x)
```
is producing 
```
def forward(self, x_1):
    empty = torch.ops.aten.empty.memory_format([3, 4], dtype = torch.float32, layout = torch.strided, device = device(type='cpu'), pin_memory = False)
    zero_ = torch.ops.aten.zero_.default(empty);  empty = None
    return zero_
```




[ghstack-poisoned]

SherlockNoMad committed:
Update on "Rewrite new_zeros, new_ones, new_full decomp with aten.full"


We should **NOT**  introducing non-functional op for decomps of functional op. 

For example 
```
make_fx(functionalize(lambda x: x.new_zeros(3)), decomposition_table=decomposition_table)(x)
```
is producing 
```
def forward(self, x_1):
    empty = torch.ops.aten.empty.memory_format([3, 4], dtype = torch.float32, layout = torch.strided, device = device(type='cpu'), pin_memory = False)
    zero_ = torch.ops.aten.zero_.default(empty);  empty = None
    return zero_
```




[ghstack-poisoned]

Lezcano commented on torch/_refs/__init__.py:
```diff
@@ -3284,12 +3308,45 @@ def new_zeros(
+    dtype = a.dtype if dtype is None else dtype
+    layout = a.layout if layout is None else layout
+    device = a.device if device is None else device
+
+    return torch.full(
+        size,
+        False,
+        dtype=dtype,
+        layout=layout,
+        device=device,
+        pin_memory=pin_memory,
+        requires_grad=requires_grad,
+    )
```
Alas, I don't think we can do this for the `new_*` functions. These functions have the property that the returned tensor is the same class as the input tensor if the input is not a `Tensor` but a subclass of it.

Now, currently, `new_empty` does not have this property either, as it dispatches to `torch.empty`. Is this an issue @ngimel @mruberry?

mruberry replied:
Interesting! I wonder if we're testing for that (if not, we should be!).

We should probably note this as an issue for now.

zou3519 replied:
> I wonder if we're testing for that (if not, we should be!).

We're not, but we definitely should. There's an issue for it over at https://github.com/pytorch/pytorch/issues/82239 .

I talked a bit with @SherlockNoMad about this offline: it's generally OK to decompose `new_zeros` into `zeros` inside the ref of an out-of-place operator, as long as the ref only contains out-of-place operations. We forgot about the case where the user is just directly calling `new_zeros` though, so this is a good catch, @Lezcano.

It seems like we want to tag our refs/decompositions with something that says when they are allowed to run. If we are running solely on regular Tensors, then many decompositions (like the new_zeros -> full) apply; if we are running on Tensor subclasses, then we cannot actually decompose a user's new_zeros call into full.

Lezcano replied:
Probably, the correct way of doing this would be to dispatch all these to either `new_full`  and implement that one correctly. I think we should try to avoid yet another way to split how a decomposition works. We are already going to have many branches (decompose at a Python level or not, decompose at an ATen level or not, decompose for backward or not, etc).

ezyang replied:
I think we should be willing to write decomps that are not subclass preserving, and just have them marked as such so they are not used in situations where you have a tensor subclass that is not a mode and you need to prop the subclass. The whole composite compliance new stuff is from the era when we didn't have modes; with modes this sort of thing is a lot less pressing

Lezcano commented on torch/_refs/__init__.py:
```diff
@@ -3275,6 +3275,30 @@ def new_empty_strided(
 
+@register_decomposition(torch.ops.aten.zeros)
+def zeros(
+    size: ShapeType,
```
This should be also a variadic argument, as `torch.zeros(2,3)` is not only a valid construction but it's the main documented overload in the docs.

SherlockNoMad committed:
Update on "Rewrite new_zeros, new_ones, new_full decomp with aten.full"


We should **NOT**  introducing non-functional op for decomps of functional op. 

For example 
```
make_fx(functionalize(lambda x: x.new_zeros(3)), decomposition_table=decomposition_table)(x)
```
is producing 
```
def forward(self, x_1):
    empty = torch.ops.aten.empty.memory_format([3, 4], dtype = torch.float32, layout = torch.strided, device = device(type='cpu'), pin_memory = False)
    zero_ = torch.ops.aten.zero_.default(empty);  empty = None
    return zero_
```




[ghstack-poisoned]

SherlockNoMad commented:
@mruberry and @Lezcano 
I am merging this PR now. Let me know if you have other comments you need to address. I can do it in another PR.

SherlockNoMad commented:
@pytorchbot merge -g

Lezcano commented:
The only open issue from my end here is https://github.com/pytorch/pytorch/pull/84946#discussion_r970514193. It'd be nice to address it in a new PR as these functions are so widely used.

ngimel commented:
And we already have `extract_shape_from_varargs` utility so that should be easy to add

SherlockNoMad commented:
> Lezcano

Fixed in https://github.com/pytorch/pytorch/pull/85117
